File: Example_ini_file.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""
    EXAMPLE FILE
    ============
    Example file for FDEM1D.

    Use:
    Calculates the forward response and sensitivity distribution for a given
    layered half-space and loop-loop configuration. Typical characteristics
    of the half-space are stored in the Model object (M) while the sensor
    characteristics are stored in the Sensor object (S). This file is
    used as example.

    Cite:
    Hanssens, D., Delefortrie, S., De Pue, J., Van Meirvenne, M., and De Smedt, P., 2018,
    Frequency domain electromagnetic 1D forward and sensitivity modelling: Overview,
    assessment and practical implementation(including MATLAB, Python and pseudo-code):
    Submitted to IEEE Geoscience and Remote Sensing Magazine.

    Code created by Daan Hanssens and Jan De Pue
    Ghent University, Belgium
    April, 2017
"""

# Import dependencies
import numpy as np
import pylab
import FDEM1D

###############################################################################
# ------------------------------------------------------- User-input -------- #
###############################################################################

#
# Sensor object (S)
#

x = 40.0                                                                        # x-coordinate receiver (m)
y = 0.0                                                                        # y-coordinate receiver (m)
z = -0.16                                                                      # z-coordinate receiver (m) - positive z-axis pointed down
height = 0.16                                                                  # Height of transmitter (m)
freq = 90e3                                                                     # Frequency (Hz)
mom = 1.0                                                                      # Transmitter moment (A.m²)
ori = 'ZZ'                                                                     # Coil orientation (-) - case sensitive

S = FDEM1D.Sensor(x, y, z, height, freq, mom, ori)                             # Sensor object


#
# Model object (M)
#

sus = np.linspace(100.0, 100.0, 100) * 1e-5                                     # Susceptibility of layer(s) (-)
con = np.linspace(.001, .001, 100)                                              # Conductivity of layer(s) (S/m)
perm = np.linspace(1.0, 1.0, 100) * 1e-12                                       # Permittivity of layer(s) (F/m)
thick = np.logspace(np.log10(.1), np.log10(.5), 100)                            # Layer(s) thickness (m)

M = FDEM1D.Model(thick, sus, con, perm)                                         # Model object


###############################################################################
# ------------------------------------------------------- Modelling --------- #
###############################################################################

# Define sensitivity parameter
par_sens = 'sus'                                                                   # Variable to be tested in sensitivity analysis

# Reflection coefficient approach (default)
[FWD_IP_RC, FWD_QP_RC] = FDEM1D.Calculate(S, M).forward()                          # Forward
[SENS_IP_RC, SENS_QP_RC, ERROR_RC] = FDEM1D.Calculate(S, M).sensitivity(par_sens)  # Sensitivity

# Propagation Matrix approach
[FWD_IP_PM, FWD_QP_PM] = FDEM1D.Calculate(S, M, method='PM').forward()                          # Forward
[SENS_IP_PM, SENS_QP_PM, ERROR_PM] = FDEM1D.Calculate(S, M, method='PM').sensitivity(par_sens)  # Sensitivity


###############################################################################
# ------------------------------------------------------- Plot -------------- #
###############################################################################

fig = pylab.figure()
ax = fig.add_subplot(211)
ax.plot(M.depth[:-1], SENS_IP_RC[:-1], '.k')
ax.plot(M.depth[:-1], SENS_IP_PM[:-1], '.r')
ax.legend(['RC', 'PM'])
ax.set_xlabel('Depth (m)')
ax.set_ylabel('IP Sensitivity: %s' % par_sens)
ax = fig.add_subplot(212)
ax.plot(M.depth[:-1], SENS_QP_RC[:-1], '.k')
ax.plot(M.depth[:-1], SENS_QP_PM[:-1], '.r')
ax.legend(['RC', 'PM'])
ax.set_xlabel('Depth (m)')
ax.set_ylabel('QP Sensitivity: %s' % par_sens)
pylab.show()

File: MATLAB/Example_ini_file.m
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%% --------------------------------------------------------------------- %%
%                              EXAMPLE FILE                               %
% ----------------------------------------------------------------------- %
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%
%  Use:
%  Calculates the forward response and sensitivity distribution for a given 
%  layered half-space and loop-loop configuration. Typical characteristics 
%  of the half-space are stored in the Model structure (M) while the sensor 
%  characteristics are stored in the Sensor structure (S). This file is
%  used as example to demonstrate the fieldnames within the structures. 
%
%  Created by Daan Hanssens
%  Ghent University, Belgium
%  December, 2016
%
%  Cite:
%  Hanssens, D., Delefortrie, S., De Pue, J., Van Meirvenne, M. and P.
%  De Smedt, Practical aspects of frequency domain electromagnetic forward 
%  and sensitivity modelling of a magnetic dipole in a multi-layered 
%  half-space: Submitted to Geoscience and Remote Sensing Magazine
%

   clc; clear all; close all;

        
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% --------------------------------------------------- User-input -------- %
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

    %
    % Sensor characteristics (S structure)
    %

        S.x=        4;                                                     % x-coordinate receiver (m)
        S.y=        0;                                                     % y-coordinate receiver (m)
        S.z=        -0.16;                                                 % z-coordinate receiver (m) - positive z-axis pointed down
        S.height=   0.16;                                                  % Height of transmitter (m)
        S.freq=     9000;                                                  % Frequency (Hz)
        S.mom=      1;                                                     % Transmitter moment (A.m^2)
        S.ori=      'ZZ';                                                  % Coil orientation (Transmitter(X,Y,Z),Receiver(X,Y,Z))


    %
    % Model characteristics (M structure)
    %

        M.sus=      linspace(100,100,100) .* 1e-5;                         % Susceptibility of layer(s) (-)
        M.con=      linspace(0.001,0.001,100);                             % Conductivity of layer(s) (S/m)
        M.perm=     linspace(1e-12,1e-12,100);                             % Permittivity of layer(s) (F/m)
        M.thick=    logspace(log10(0.1),log10(0.5),100);                   % Layer(s) thickness (m)


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% --------------------------------------------------- Modelling --------- %
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

    %
    % Calculate forward response (ppm)
    %

        [FWD_IP,FWD_QP] = FDEM1DFWD_RC(S,M);                               % Reflection coefficient approach
%        [FWD_IP,FWD_QP] = FDEM1DFWD_PM(S,M);                              % Propagation matrix approach


    %
    % Calculate sensitivity response (ppm)
    %
    
        par= 'con';  % ('con','sus','perm')
        [SENS_IP,SENS_QP,Err] = FDEM1DSENS_RC(S,M,par);                    % Reflection coefficient approach
%         [SENS_IP,SENS_QP,Err] = FDEM1DSENS_PM(S,M,par);                  % Propagation matrix approach
        
        
        %
        % Plot sensitivity
        % (Optionally: exclude basement layer due to increased sensitivity 
        % of infinite basement layer)
        %

            figure(); 
            subplot(2,1,1);
            plot(cumsum(M.thick(1:end-1)),SENS_QP(1:end-1),'k.'); 
            legend(['Error < ',num2str(abs(Err.QP))]); legend boxoff;
            ylabel(['Sensitivity: QP, ',par]); 
            subplot(2,1,2);
            plot(cumsum(M.thick(1:end-1)),SENS_IP(1:end-1),'k.'); 
            legend(['Error < ',num2str(abs(Err.IP))]); legend boxoff;
            xlabel('Depth (m)'); ylabel(['Sensitivity: IP, ',par]);
    
    
